src/backend/libinput.rs

```diff
@@ -0,0 +1,147 @@
+//! Implementation of input backend trait for types provided by `libinput`
+
+use backend::SeatInternal;
+use backend::input::{InputBackend, InputHandler, Seat, SeatCapabilities};
+use input::{Libinput, Device, Seat as LibinputSeat, DeviceCapability};
+use input::event::*;
+
+use std::io::Error as IoError;
+use std::collections::hash_map::{DefaultHasher, HashMap};
+use std::hash::{Hash, Hasher};
+
+pub struct LibinputInputBackend {
+    context: Libinput,
+    devices: Vec<Device>,
+    seats: HashMap<LibinputSeat, Seat>,
+    handler: Option<Box<InputHandler<LibinputInputBackend> + 'static>>,
+}
+
+impl InputBackend for LibinputInputBackend {
+    type InputConfig = [Device];
+    type EventError = IoError;
+
+    fn set_handler<H: InputHandler<Self> + 'static>(&mut self, mut handler: H) {
+        if self.handler.is_some() {
+            self.clear_handler();
+        }
+        for seat in self.seats.values() {
+            handler.on_seat_created(&seat);
+        }
+        self.handler = Some(Box::new(handler));
+    }
+
+    fn get_handler(&mut self) -> Option<&mut InputHandler<Self>> {
+        self.handler
+            .as_mut()
+            .map(|handler| handler as &mut InputHandler<Self>)
+    }
+
+    fn clear_handler(&mut self) {
+        if let Some(mut handler) = self.handler.take() {
+            for seat in self.seats.values() {
+                handler.on_seat_destroyed(&seat);
+            }
+        }
+    }
+
+    fn input_config(&mut self) -> &mut Self::InputConfig {
+        &mut self.devices
+    }
+
+    fn set_cursor_position(&mut self, _x: u32, _y: u32) -> Result<(), ()> {
+        // FIXME later.
+        // This will be doable with the hardware cursor api and probably some more cases
+        Err(())
+    }
+
+    fn dispatch_new_events(&mut self) -> Result<(), IoError> {
+        self.context.dispatch()?;
+        for event in &mut self.context {
+            match event {
+                Event::Device(device_event) => {
+                    use input::event::device::*;
+                    match device_event {
+                        DeviceEvent::Added(device_added_event) => {
+                            let added = device_added_event.into_event().device();
+
+                            let new_caps = SeatCapabilities {
+                                pointer: added.has_capability(DeviceCapability::Pointer),
+                                keyboard: added.has_capability(DeviceCapability::Keyboard),
+                                touch: added.has_capability(DeviceCapability::Touch),
+                            };
+
+                            let device_seat = added.seat();
+                            self.devices.push(added);
+
+                            let contains = self.seats.contains_key(&device_seat);
```

Comment:
Good idea, but the thing is I also need a mutable reference to `self.seats` in line 91.

Otherwise I could also have done `if let Some(old_seat) = self.seats.get_mut(&device_seat) {` there.

Maybe I am missing something, but I don't think the Entry API will help here?